### PR TITLE
bitvector representation cleanup

### PR DESCRIPTION
**Pull request by kroening**
These are the parts of #3107 that do not change behavior.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/util/simplify_expr_int.cpp [hunk 1]
```diff
@@ -1269,9 +1269,11 @@ bool simplify_exprt::simplify_bitnot(exprt &expr)
...
-  if(expr.type().id()==ID_bv ||
-     expr.type().id()==ID_unsignedbv ||
-     expr.type().id()==ID_signedbv)
+  const auto &type = expr.type();
+
+  if(
+    type.id() == ID_bv || type.id() == ID_unsignedbv ||
+    type.id() == ID_signedbv)
   {
     const auto width = to_bitvector_type(expr.type()).get_width();
```
Comment (tautschnig) [inline src/util/simplify_expr_int.cpp]: If the above cleanup is then, then surely `type` should be used here as well?

File: src/util/simplify_expr_int.cpp [hunk 2]
```diff
@@ -885,8 +885,9 @@ bool simplify_exprt::simplify_concatenation(exprt &expr)
       exprt &op=*it;
       if(op.is_true() || op.is_false())
       {
-        bool value=op.is_true();
-        op=constant_exprt(value?ID_1:ID_0, unsignedbv_typet(1));
+        const bool value = op.is_true();
+        op = from_integer(value, unsignedbv_typet(1));
+        result = true;
```
Comment (tautschnig) [inline src/util/simplify_expr_int.cpp]: I may be confused: there has been a modification, hence the result should be `false`?!

**Review by tautschnig — CHANGES_REQUESTED** [review]
A first few comments below, but also this seems to spuriously adjust the submodule pointer for the java-models-library, which seems wrong.

File: src/util/simplify_expr_int.cpp (hunk 2 above)
Comment (kroening) [reply src/util/simplify_expr_int.cpp]: Indeed, fixed

File: src/util/simplify_expr_int.cpp (hunk 1 above)
Comment (kroening) [reply src/util/simplify_expr_int.cpp]: Now cleaned that as well.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 94086ac5317fe840e913c4354961d8db37760855).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88403914

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 94e86296e3354bd2fa97dc707538705b5a61889d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88427002

**Review by tautschnig — APPROVED** [review]
The `simplify_concatenation` bug fixes would likely deserve a note in the commit message (and they are also part of one of my PRs that I need to get back to ...), but otherwise now does what is says on the tin.

**Comment by kroening** [comment]
@tautschnig I've tweaked the commit message.

File: src/util/simplify_expr_int.cpp [hunk 3]
```diff
@@ -787,10 +787,12 @@ bool simplify_exprt::simplify_bitwise(exprt &expr)
...
     {
-      if(it->is_constant() &&
-         id2string(to_constant_expr(*it).get_value()).find('0')==
-           std::string::npos &&
-         expr.operands().size()>1)
+      if(
+        it->is_constant() &&
+        bv2integer(
+          id2string(to_constant_expr(*it).get_value()), width, false) ==
+          power(2, width) - 1 &&
```
Comment (romainbrenguier) [inline src/util/simplify_expr_int.cpp:794]: It's not obvious what this condition means. Defining an intermediary variable would have helped.

File: src/util/simplify_expr.cpp [hunk 4]
```diff
@@ -1567,14 +1567,15 @@ optionalt<std::string> simplify_exprt::expr2bits(
        type.id()==ID_floatbv ||
        type.id()==ID_fixedbv)
     {
-      std::string nat=id2string(to_constant_expr(expr).get_value());
-      std::reverse(nat.begin(), nat.end());
+      const auto width = to_bitvector_type(type).get_width();
+      const auto &bvrep = to_constant_expr(expr).get_value();
 
       endianness_mapt map(type, little_endian, ns);
```
Comment (romainbrenguier) [inline src/util/simplify_expr.cpp:1575]: map could be const

File: src/util/simplify_expr.cpp [hunk 5]
```diff
@@ -1458,7 +1458,9 @@ exprt simplify_exprt::bits2expr(
       tmp[i]=bits[map.map_bit(i)];
 
     std::reverse(tmp.begin(), tmp.end());
-    return constant_exprt(tmp, type);
+
+    mp_integer i = binary2integer(tmp, false);
```
Comment (romainbrenguier) [inline src/util/simplify_expr.cpp:1462]: const

File: src/solvers/lowering/popcount.cpp [hunk 6]
```diff
@@ -47,7 +47,8 @@ exprt lower_popcount(const popcount_exprt &expr, const namespacet &ns)
     bitstring.reserve(new_width);
     for(std::size_t i = 0; i < new_width / (2 * shift); ++i)
       bitstring += std::string(shift, '0') + std::string(shift, '1');
-    constant_exprt bitmask(bitstring, x.type());
+    const mp_integer value = binary2integer(bitstring, false);
+    constant_exprt bitmask(integer2bv(value, new_width), x.type());
```
Comment (romainbrenguier) [inline src/solvers/lowering/popcount.cpp:51]: const

File: src/solvers/refinement/string_refinement.cpp [hunk 7]
```diff
@@ -1449,8 +1449,8 @@ static exprt sum_over_map(
     int second=negated?(-term.second):term.second;
     if(t.id()==ID_constant)
     {
-      std::string value(to_constant_expr(t).get_value().c_str());
-      constants+=binary2integer(value, true)*second;
+      const auto int_value = numeric_cast_v<mp_integer>(to_constant_expr(t));
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_refinement.cpp:1452]: `to_constant_expr` shouldn't be required (implicitly done in numeric_cast_v)

**Review by romainbrenguier — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: e5850b8cb7dfea51c5579832411823d2973f2dab).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88518428

**Pull request merged**